File: backend/src/routes/latest.ts
import { createRoute, OpenAPIHono, z } from "@hono/zod-openapi";
import { db } from "../db.js";
import { measurements, type UnitAnalysis } from "../schema.js";
import { eq, desc } from "drizzle-orm";
import { InstallationIdSchema, LatestMeasurementSchema } from "../validation.js";

const app = new OpenAPIHono();

const latestRoute = createRoute({
  method: "get",
  path: "/{installation_id}",
  request: {
    params: z.object({
      installation_id: InstallationIdSchema,
    }),
  },
  responses: {
    200: {
      content: {
        "application/json": {
          schema: LatestMeasurementSchema,
        },
      },
      description: "Latest measurement",
    },
    404: {
      description: "No measurements found",
    },
  },
});

app.openapi(latestRoute, async (c) => {
  const { installation_id: installationId } = c.req.valid("param");

  const latest = await db.query.measurements.findFirst({
    where: eq(measurements.installationId, installationId),
    orderBy: [desc(measurements.capturedAt)],
  });

  if (!latest) {
    return c.json({ error: "No measurements found for this installation" }, 404);
  }

  return c.json({
    installation_id: latest.installationId,
    captured_at: latest.capturedAt.toISOString(),
    pushed_at: latest.pushedAt?.toISOString() ?? null,
    chlorine: {
      status: latest.chlorineStatus as UnitAnalysis["status"],
      diagnosis: latest.chlorineDiagnosis,
      pattern_detected: latest.chlorinePattern,
      blinking_leds: latest.chlorineBlinking,
      solid_leds: latest.chlorineSolid,
      summary: latest.chlorineSummary ?? "",
      action_required: latest.chlorineAction,
      recommended_action: latest.chlorineRecommended ?? "",
    },
    ph: {
      status: latest.phStatus as UnitAnalysis["status"],
      diagnosis: latest.phDiagnosis,
      pattern_detected: latest.phPattern,
      blinking_leds: latest.phBlinking,
      solid_leds: latest.phSolid,
      summary: latest.phSummary ?? "",
      action_required: latest.phAction,
      recommended_action: latest.phRecommended ?? "",
    },
    raw_response: latest.rawResponse,
  });
});

export default app;

File: backend/src/routes/latest.py
from auth import verify_token
from db.models import Installation, Measurement
from db.session import get_db
from fastapi import APIRouter, Depends, HTTPException
from measurement_service import latest_schema_from_measurement
from schemas.models import LatestMeasurementSchema, validate_installation_id
from sqlalchemy import desc
from sqlalchemy.orm import Session

router = APIRouter()


@router.get("/{installation_id}", response_model=LatestMeasurementSchema)
async def get_latest_measurement(
    installation_id: str,
    db: Session = Depends(get_db),
    _auth: None = Depends(verify_token),
) -> LatestMeasurementSchema:
    try:
        validate_installation_id(installation_id)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e)) from e

    latest = (
        db.query(Measurement)
        .filter(Measurement.installation_id == installation_id)
        .order_by(desc(Measurement.captured_at))
        .first()
    )

    if not latest:
        raise HTTPException(
            status_code=404, detail="No measurements found for this installation"
        )

    installation = db.get(Installation, installation_id)
    sensors = installation.shared_sensors if installation else []

    return latest_schema_from_measurement(latest, sensors)
